File: pkg/client/http.go
package client

import (
	"io"
	"net/http"
	"net/http/cookiejar"
	"strings"
	"time"
)

type realHTTPClient struct {
	client *http.Client
}

func NewHTTPClient() HTTPClient {
	jar, _ := cookiejar.New(nil)
	return &realHTTPClient{
		client: &http.Client{
			Jar:     jar,
			Timeout: 5 * time.Second,
		},
	}
}

func (h *realHTTPClient) Get(url string, headers map[string]string) HTTPResponse {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return HTTPResponse{Err: err}
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	return h.do(req)
}

func (h *realHTTPClient) Post(url string, data string, headers map[string]string) HTTPResponse {
	var body io.Reader
	if data != "" {
		body = strings.NewReader(data)
	}
	req, err := http.NewRequest("POST", url, body)
	if err != nil {
		return HTTPResponse{Err: err}
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	return h.do(req)
}

func (h *realHTTPClient) do(req *http.Request) HTTPResponse {
	resp, err := h.client.Do(req)
	if err != nil {
		return HTTPResponse{Err: err}
	}
	defer resp.Body.Close()
	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return HTTPResponse{StatusCode: resp.StatusCode, Err: err}
	}
	return HTTPResponse{StatusCode: resp.StatusCode, Body: string(b)}
}

File: tl_mr6400/http.py
import urllib.request
import http.cookiejar


class HttpResponse:
    def __init__(self, status_code: int, text: str):
        self.status_code = status_code
        self.text = text


class HttpSession:
    def __init__(self, timeout: int = 5):
        self.timeout = timeout
        cookie_jar = http.cookiejar.CookieJar()
        self._opener = urllib.request.build_opener(
            urllib.request.HTTPCookieProcessor(cookie_jar)
        )

    def get(self, url: str, headers: dict = None, **kwargs) -> HttpResponse:
        return self._request(url, headers=headers, **kwargs)

    def post(self, url: str, data: str = None, headers: dict = None, **kwargs) -> HttpResponse:
        return self._request(url, data=data, headers=headers, **kwargs)

    def _request(self, url: str, data: str = None, headers: dict = None, **kwargs) -> HttpResponse:
        body = data.encode("utf-8") if data else None
        req = urllib.request.Request(url, data=body)
        for key, val in (headers or {}).items():
            req.add_header(key, val)

        timeout = kwargs.get("timeout", self.timeout)
        try:
            resp = self._opener.open(req, timeout=timeout)
            return HttpResponse(resp.status, resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            return HttpResponse(e.code, e.read().decode("utf-8"))
